### PR TITLE
Optimize check owned by removing json and fix double rkyv serialization bug

### LINKED ISSUE #113: Double rkyv serialization in compat/rkyv.rs
here were rkyv serialize once before we do it agian when we pass it to rkyv_into_ptr https://github.com/dusk-network/wallet-core/blob/main/src/compat/rkyv.rs#L100

### PR DESCRIPTION
**Pull request by Daksh14**
Closes #113 and #114

This is the wasm binary that is in the latest dusk-wallet-js version, we should merge into main ASAP

### CODE REVIEW

File: src/compat/crypto.rs [hunk 1]
```diff
@@ -4,73 +4,104 @@
...
+                    return utils::fail();
+                }
+            };
 
-            break;
+        last_pos = core::cmp::max(last_pos, *note.pos());
+
+        for idx in 0..=MAX_KEY {
+            let idx = idx as u64;
+            let view_key = key::derive_vk(&seed, idx);
```
Comment (ZER0) [inline src/compat/crypto.rs:63]: Let's create a vector of keys, and let's derive them once for all. We don't need to do every time the same operation, over and over.
Here you're deriving the same keys every time you're comparing a note, and we're talking about not only unnecessary operation, but performing unnecessary scalar multiplications, that we know are the bottleneck in our code.

My suggestion is having an array (since `MAX_KEY` is constant, right?) of `view_keys` and one of `secret_keys` that we instantiate at the beginning, before any loop.

File: src/compat/crypto.rs [hunk 2]
```diff
@@ -4,73 +4,104 @@
...
 
-        if vk.owns(&note) {
-            let nullifier = note.gen_nullifier(&sk);
-
-            nullifier_found = nullifier;
-            is_owned = true;
-            pk_found = Some(PublicKey::from(&sk));
+    for leaf_bytes in leaf_chunk.by_ref() {
+        let TreeLeaf { block_height, note } =
+            match rkyv::from_bytes(leaf_bytes).ok() {
```
Comment (ZER0) [inline src/compat/crypto.rs:52]: Is there any reason why you transform a `Result` to an `Option` instead of checking directly the `Result`?

I saw you're doing that in multiple places, not going to comment all of those, just reply here.

File: src/compat/crypto.rs [hunk 3]
```diff
@@ -4,73 +4,104 @@
...
+
+                let psk_found =
+                    bs58::encode(PublicKey::from(sk).to_bytes()).into_string();
+
+                let raw_note: Vec<u8> =
+                    match rkyv::to_bytes::<Note, MAX_LEN>(&note).ok() {
+                        Some(n) => n.to_vec(),
+                        None => return utils::fail(),
+                    };
+
```
Comment (ZER0) [inline src/compat/crypto.rs:84]: We will need to change the ergonomics of how to handle the failure, soon or later.

File: src/compat/crypto.rs [hunk 4]
```diff
@@ -4,73 +4,104 @@
...
 
     utils::into_ptr(types::CheckNoteOwnershipResponse {
-        is_owned,
-        nullifier: nullifier_found,
-        public_key: pk_found,
+        notes,
+        block_heights,
+        public_spend_keys,
+        nullifiers,
+        last_pos,
```
Comment (ZER0) [inline src/compat/crypto.rs:104]: We also need to redo this. Not in this PR obviously, no worries. :)

But the fact that we're still using strings (the list of block heights as string is just plain wrong, we should have them as vec of numbers or bytes), it still implies some form of additional serialization.

File: src/compat/crypto.rs [hunk 5]
```diff
@@ -4,73 +4,104 @@
...
 #[no_mangle]
 pub fn check_note_ownership(args: i32, len: i32) -> i64 {
-    // we just use BalanceArgs again as we don't want to add more cluter types
-    // when the data you want is the same
-    let types::CheckNoteOwnershipArgs { note, seed } =
-        match utils::take_args(args, len) {
-            Some(a) => a,
-            None => return utils::fail(),
-        };
+    let args = utils::take_args_raw(args, len);
```
Comment (ZER0) [inline src/compat/crypto.rs:32]: Why don't use `from_raw_parts` directly? The overhead to have an additional function call is not worth, there really nothing more than that. If you need the cast, you can either have `.into()` where supported or `as` directly in the `from_raw_parts` call's arguments.

Plus, you're wrongly discard an `unsafe` keyword where is absolutely necessary: here I might be tempted to believe you actually did a safe interface to an `unsafe` call, but it's not the case: you're not performing any checks or doing anything that would imply such thing.

File: tests/wallet.rs [hunk 6]
```diff
@@ -336,7 +336,8 @@ mod node {
             .into_iter()
             .map(|value| {
                 let obfuscated = (rng.next_u32() & 1) == 1;
-                let idx = rng.next_u64() % MAX_KEY as u64;
+                let idx = rng.next_u64()
+                    % (if MAX_KEY == 0 { 1 } else { MAX_KEY }) as u64;
```
Comment (ZER0) [inline tests/wallet.rs:340]: This check is useless, `MAX_KEY` should never be `0`, if someone set `0` by mistake it's totally fine that the test would fail because of that. Please, put back the previous code.

**Review by ZER0 — CHANGES_REQUESTED** [review]
Overall looks good, but there are few nits to change, one very important tied to performance and scalar multiplication.

File: src/compat/crypto.rs (hunk 2 above)
Comment (Daksh14) [reply src/compat/crypto.rs]: I will change them in a later PR

File: Cargo.toml [hunk 7]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "dusk-wallet-core"
-version = "0.21.6"
+version = "0.21.8"
```
[the remark this answers is not in the document]
Comment (Daksh14) [reply Cargo.toml:3]: its published but it wasn't incremented in the cargo.toml but the npm package registry instead but we ended up embeding it in the wallet-js

File: src/compat/crypto.rs [hunk 8]
```diff
@@ -4,73 +4,113 @@
...
 
-        if vk.owns(&note) {
-            let nullifier = note.gen_nullifier(&sk);
+    for idx in 0..MAX_KEY {
+        let idx = idx as u64;
+        let view_key = key::derive_vk(&seed, idx);
+        let sk = key::derive_sk(&seed, idx as _);
+        view_keys.push(view_key);
+        secret_keys.push(sk);
+    }
```
Comment (ZER0) [inline src/compat/crypto.rs:60]: As mentioned during our pairing, my suggestion was using arrays, not vectors, since you know the size and you won't need to make them mutable. So this and the loop above can be replaced by:
```rust
    let view_keys: [ViewKey; MAX_KEY] =
        core::array::from_fn(|i| key::derive_vk(&seed, i as _));
    let secret_keys: [SecretKey; MAX_KEY] = 
        core::array::from_fn(|i| key::derive_sk(&seed, i as _));
```

It's more concise and you won't need dynamic allocation for the vectors neither mutable variable since you populate the arrays immediately when you instance them.

File: src/compat/crypto.rs [hunk 9]
```diff
@@ -4,73 +4,113 @@
...
-            is_owned = true;
-            pk_found = Some(PublicKey::from(&sk));
+    for leaf_bytes in leaf_chunk.by_ref() {
+        let TreeLeaf { block_height, note } = match rkyv::from_bytes(leaf_bytes)
+        {
+            Ok(a) => a,
+            Err(_) => {
+                return utils::fail();
+            }
+        };
```
Comment (ZER0) [inline src/compat/crypto.rs:69]: BTW, this was the reason why we created the simple macro `unwrap_or_bail`, that we definitely are going to reintroduce once we refactor `wallet-core` and manage the error in a better way (not in this PR, ofc)

File: src/compat/crypto.rs [hunk 10]
```diff
@@ -4,73 +4,113 @@
...
+            if view_keys[idx].owns(&note) {
+                let sk = secret_keys[idx];
+                let nullifier = note.gen_nullifier(&sk);
+
+                let nullifier_found =
+                    match rkyv::to_bytes::<BlsScalar, MAX_LEN>(&nullifier).ok()
+                    {
+                        Some(n) => n.to_vec(),
+                        None => return utils::fail(),
+                    };
```
Comment (ZER0) [inline src/compat/crypto.rs:83]: This is the same code you had before and I asked you to replace in the whole PR: you don't need the `.ok()` method and then match the `Option`, you should be able to directly match the `Result`.

File: src/ffi.rs [hunk 11]
```diff
@@ -92,15 +92,15 @@ pub fn balance(args: i32, len: i32) -> i64 {
         Err(_) => return utils::fail(),
     };
 
-    let mut keys = unsafe { [mem::zeroed(); MAX_KEY + 1] };
+    let mut keys = unsafe { [mem::zeroed(); MAX_KEY] };
```
Comment (ZER0) [inline src/ffi.rs:95]: You can use the same approach I mentioned in the previous comment.

File: src/ffi.rs [hunk 12]
```diff
@@ -92,15 +92,15 @@ pub fn balance(args: i32, len: i32) -> i64 {
...
     let mut sum = 0u64;
 
     'outer: for note in notes {
         // we iterate all the available keys until one can successfully decrypt
         // the note. if all fails, returns false
-        for idx in 0..=MAX_KEY {
+        for idx in 0..MAX_KEY {
             if keys_len == idx {
                 keys[idx] = key::derive_vk(&seed, idx as u64);
                 keys_len += 1;
```
Comment (ZER0) [inline src/ffi.rs:106]: What's `keys_len` used for?

File: src/ffi.rs [hunk 13]
```diff
@@ -92,15 +92,15 @@ pub fn balance(args: i32, len: i32) -> i64 {
...
     let mut keys_len = 0;
     let mut sum = 0u64;
 
     'outer: for note in notes {
         // we iterate all the available keys until one can successfully decrypt
         // the note. if all fails, returns false
-        for idx in 0..=MAX_KEY {
+        for idx in 0..MAX_KEY {
             if keys_len == idx {
                 keys[idx] = key::derive_vk(&seed, idx as u64);
```
Comment (ZER0) [inline src/ffi.rs:105]: Also here, you shouldn't `derive` the keys in the loop, for the same reason we discussed above. You should just instatiate outside as we did above.

File: src/ffi.rs [hunk 14]
```diff
@@ -355,7 +355,7 @@ pub fn public_keys(args: i32, len: i32) -> i64 {
         None => return utils::fail(),
     };
 
-    let keys = (0..=MAX_KEY)
+    let keys = (0..MAX_KEY)
         .map(|idx| key::derive_pk(&seed, idx as u64))
```
Comment (ZER0) [inline src/ffi.rs:359]: Same here. Just to be sure, check if we have any part of the code where this pattern is repeated, and replace it with the array outside the loop.

File: src/ffi.rs [hunk 15]
```diff
@@ -382,7 +382,7 @@ pub fn view_keys(args: i32, len: i32) -> i64 {
         None => return utils::fail(),
     };
 
-    let keys: Vec<_> = (0..=MAX_KEY)
+    let keys: Vec<_> = (0..MAX_KEY)
```
Comment (ZER0) [inline src/ffi.rs:385]: You don't need it if you have already the array as mentioned. That would reduce another loop.

File: src/ffi.rs [hunk 16]
```diff
@@ -416,14 +416,14 @@ pub fn nullifiers(args: i32, len: i32) -> i64 {
     };
 
     let mut nullifiers = Vec::with_capacity(notes.len());
-    let mut sks = unsafe { [mem::zeroed(); MAX_KEY + 1] };
-    let mut vks = unsafe { [mem::zeroed(); MAX_KEY + 1] };
+    let mut sks = unsafe { [mem::zeroed(); MAX_KEY] };
+    let mut vks = unsafe { [mem::zeroed(); MAX_KEY] };
```
Comment (ZER0) [inline src/ffi.rs:420]: Same here.

File: src/ffi.rs [hunk 17]
```diff
@@ -416,14 +416,14 @@ pub fn nullifiers(args: i32, len: i32) -> i64 {
...
     let mut keys_len = 0;
 
     'outer: for note in notes {
         // we iterate all the available view key until one can successfully
         // decrypt the note. if any fails, returns false
-        for idx in 0..=MAX_KEY {
+        for idx in 0..MAX_KEY {
```
Comment (ZER0) [inline src/ffi.rs:426]: And here.

File: tests/wallet.rs [hunk 18]
```diff
@@ -336,7 +336,7 @@ mod node {
             .into_iter()
             .map(|value| {
                 let obfuscated = (rng.next_u32() & 1) == 1;
-                let idx = rng.next_u64() % MAX_KEY as u64;
+                let idx = rng.next_u64() % (MAX_KEY) as u64;
```
Comment (ZER0) [inline tests/wallet.rs:339]: You don't need to wrap the `MAX_KEY` in parentheses, the previous code would works fine.

**Review by ZER0 — CHANGES_REQUESTED** [review]
Still few things to change, but we're on the right track. Be sure to do not derive the keys inside any loops, but doing so upfront. If we notice this as common pattern, (the need of `sk` and `vk` together), we might also have an inline function that does that for us, so we don't have to repeat it.

File: src/compat/crypto.rs (hunk 9 above)
Comment (Daksh14) [reply src/compat/crypto.rs:62]: https://github.com/dusk-network/wallet-core/issues/117

File: src/ffi.rs [hunk 19]
```diff
@@ -416,20 +412,16 @@ pub fn nullifiers(args: i32, len: i32) -> i64 {
...
-    let mut sks = unsafe { [mem::zeroed(); MAX_KEY + 1] };
-    let mut vks = unsafe { [mem::zeroed(); MAX_KEY + 1] };
-    let mut keys_len = 0;
+
+    let sks: [SecretKey; MAX_KEY] =
+        core::array::from_fn(|i| key::derive_sk(&seed, i as _));
+    let vks: [ViewKey; MAX_KEY] =
+        core::array::from_fn(|i| key::derive_vk(&seed, i as _));
 
     'outer: for note in notes {
```
Comment (ZER0) [inline src/ffi.rs:421]: You still didn't modify this loop as discussed.

1. The comment is wrong, or best case scenario misleading
2. The usage of labels makes the flow harder to follow from top to bottom – it's "GOTO-like" –; therefore the labels should be used only when there is not any decent alternative available
3. Due to the label usage, the code makes use of a temp var called `keys_len` that shouldn't be there.

All those should be fixed, if you don't want to do it in this PR feel free to open immediately a new one – but also an issue related to that – to apply those changes. I suggested to do it in this PR since you're already touching this code and in addition you can skip to create a separate issue. But if you prefer to file a new issue and fix it immediately in a new PR, it's also fine.

My suggestion, as I already mentioned to you, would be modify this using a more functional approach with `try_fold`. Here is the code:

```rust
    let nullifiers = notes.iter().try_fold(
        Vec::with_capacity(notes.len()),
        |mut nullifiers, note| {
            if let Some(idx) = (0..MAX_KEY).find(|&i| vks[i].owns(note)) {
                let nullifier = note.gen_nullifier(&sks[idx]);
                nullifiers.push(nullifier);
                Ok(nullifiers)
            } else {
                Err(())
            }
        },
    );

    match nullifiers {
        Ok(n) => utils::rkyv_into_ptr(n),
        Err(_) => utils::fail(),
    }
```

Basically, if you cannot decrypt a note with *any* of the keys you own, than the function calls `utils::fail`.

**Review by ZER0 — CHANGES_REQUESTED** [review]
There is the part related to the loops and labels usage (see the comment). Once it's fixed I will approve immediately (that loop and others like that).

I would suggest to fix it in this PR to avoid open a separate issue, but if you're more comfortable to file a new issue and fix it there immediately please file the issue and add its link as reply to the conversation in this review concerning the matter; I'm also fine with that as soon as it's fixed ASAP.

File: src/ffi.rs (hunk 19 above)
Comment (Daksh14) [reply src/ffi.rs:421]: I removed the comment and the label, I think any greater refractor than that would be counter productive since the "big" refractor is imminent, I have already created the EPIC for refractor I'll add this content to it 
https://github.com/dusk-network/wallet-core/issues/117

File: src/ffi.rs [hunk 20]
```diff
@@ -416,27 +412,18 @@ pub fn nullifiers(args: i32, len: i32) -> i64 {
...
+    let vks: [ViewKey; MAX_KEY] =
+        core::array::from_fn(|i| key::derive_vk(&seed, i as _));
 
-            if vks[idx].owns(&note) {
-                nullifiers.push(note.gen_nullifier(&sks[idx]));
-                continue 'outer;
-            }
-        }
+    for note in notes {
+        let Some(vk_idx) = vks.iter().position(|vk| vk.owns(&note)) else {
```
Comment (ZER0) [inline src/ffi.rs:422]: I forgot about `position`, this is much better. Just call it `idx` or `i`.

**Review by ZER0 — APPROVED** [review]
LGTM, just few minor nits before merging. See the comments, but to summarize:

1. rename `vk_idx` into `idx`
2. remove the unused `core::mem` as stated by github too.
3. remove the sorrounding parentheses from `MAX_KEY` from tests.
